design/pool-configuration.md

```diff
@@ -0,0 +1,533 @@
+apiVersion: metallb.io/v1beta1
+kind: L2Advertisement
+metadata:
+  name: lsadvertisement
+  namespace: metallb-system
+spec:
+    pools:
+      - pool1
+      - pool2
+    nodeSelectors:
+    - matchExpressions:
+      - key: kubernetes.io/hostname
+        operator: In
+        values: [hostA, hostB]
+```
+
+#### Announcing the IP to a subset of BGP neighbors
+
+This concern is now addressed by adding a peers selector to the `BGPAdvertisement` entity.
+By using both the pools and the peers selector, it will be possible to declare the intent of announcing the IPs of a given `AddressPool` only to a subset of BGP peers,
+and only from a subset of nodes.
+
+```go
+type bgpAdvertisement struct {
+    Name              string
+  Pools             []string
+  Peers             []string
+  PoolSelector      map[string]string
+  PeerSelector      map[string]string
+  AggregationLength *int
+  LocalPref         *uint32
+  Communities       []string
+}
+```
+
+Example CRD:
+
+```yaml
+apiVersion: metallb.io/v1beta1
+kGPd: BGPAdvertisement
+metadata:
+  name: bgpadvertisement
+  namespace: metallb-system
+spec:
+    pools:
+      - pool1
+      - pool2
+    peers:
+      - peer1
+      - peer2
+    - communities:
+        - 65535:65282
+    aggregationLength: 32
+    localPref: 100
+```
+
+#### Not announcing the IP
+
+This concern will be addressed by defining an AddressPool not referenced by any Advertisement instance.
+
+#### Using a given AddressPool only for a subset of services
+
+The idea is to be able to reduce the scope of a particular AddressPool to a set of namespaces and services, by adding an optional namespace / service selector.
+
+```go
+type addressPool struct {
+  Name              string
+  Addresses         []string
+  AvoidBuggyIPs     bool
+  AutoAssign        *bool
+  AllocateTo        serviceAllocation
+}
+
+type serviceAllocation struct {
+  Priority          int
+  Namespaces        []string
+  NamespaceSelector map[string]string
+  ServiceSelector   map[string]string
+}
+```
+
+Example CRD
+
+```yaml
+apiVersion: metallb.io/v1beta1
+kind: AddressPool
+metadata:
+  name: addresspool-sample3
+  namespace: metallb-system
+spec:
+  addresses:
+    - 2002:2:2::1-2002:2:2::100
+    - 172.20.0.100/24
+  autoAssign: false
+  avoidBuggyIPs: true
+  allocateTo:
+    priority: 50
+      namespaces:
+        - namespaceA
+        - namespaceB
+      serviceSelector:
+          tenant: foo
+```
+
+The controller will use the pool with highest priority in case of multiple matches. This is deterministic and will make the behaviour more clear to the users.
+
+Open question: what to do in case a selective pool finishes the IPs? As a user, if I choose a given pool from my services, I'd rather not have the IP assigned than getting one
+from another pool. One alternative is to keep this as default (i.e. if allocate-to is set and no IPs are available, do not go to the next pool). Another option is to have another
+flag that overrides this behaviour.
```

Comment:
I think if allocate-to is set and there are no IPs available AND there are no other pools referenced in the {BGP,L2}Advertisement, I'd go with no have the IP assigned. I think I prefer that over the flag. But no strong feelings if you think otherwise.

Another option might be to make the edit of an `AddressPool` fail if they remove an IP that is currently allocated? In that way, they can create a new service (or delete and re-create) even with a specific IP selected, then remove this service and edit the AddressPool (this time won't fail, as the IP to remove is not allocated).

Maybe that is better? What do you think?

Reply:
The problem I want to address here is, what if we have two valid AddressPools, one with the selector and the priority field set, the other one with just the selector (or the selector but a lower priority).
In my opinion we must just fail, because the behaviour is predictable and the user expressed his willing to use a specific address pool for a given service. Otherwise it would be harder for them to understand how MetalLB is behaving.

I'll go with the failure approach.

Reply:
Priority seems to be mandatory (it is an int, not an *int) so if the selector is present, a priority must be used (default 0?), right? I assume, then, it is the case with a lower priority you mentioned and not that other one.

Assuming that case, I'm not sure I'm missing something but just failing seems weird. LEt's say you have a namespace and 3 different address pools to use there. You _need_ to set a priority on all. Let's assume no two pool have the same priority (we would later need to address if having same prio is allowed or not), so you assign them all to one namespace.

The pool with highest prio runs out of IPs. Then, no new service will have an IP allocated. So, what is the point of having more address pools that are assigned to the namespace?

I guess it only makes sense to have several address pools assigned to one namespace if we also allow the prio to be the same. I don't think we should check any of this, just document whenever we write user-facing docs.

Or am I missing something @fedepaol ?

Reply:
If that makes sense and we allow several address pool to use the same prio, then I think that it can work.

It seems it can be tricky the algorithm to decide which AddressPool to use to allocate when a new service is created. But I trust you have thought that too :)

Reply:
Priority has to be optional, because of the usual "keep it simple as long as we don't need complexity" logic.
What I am worried about here is a scenario where a user has some "special" pools for a namespace, and a bunch of other "regular" pools with no priority.

What I want to avoid is having the services in the special namespace pick an ip from the "regular" pools, that's why I am suggesting to fail.
But you are probably right, we can just document it and users will understand they had to put the right priorities and ranges in place. In that case, it will be a matter of changing the current logic to sort the pools by priority.

Reply:
@fedepaol I mean, from the golang POV if that field is not a pointer, it will be 0 when not present. 0 might be a reasonable default, though.

I'm kind of confused, though. I'm sure I'm missing something. But if we add a priority field and only allocate from the pool with highest prio, what is the point of having the other address pools matching the same selector with lower prio?

I mean, I don't follow why we have a prio field if basically setting it means that only the higher prio will be used and the rest ignored (seems like an easy way to ignore the rest of the pools).

What am I missing?

Reply:
The point is, not having a selector currently means "select everything" (and we want to keep that).

With that in mind, you can have some "general pools" good for everything, and one namespace specific pool.
Also, the selector is a label selector, so it could be that we have a broader selector that also impacts (as a side effect) the "finer selected" pools.

An example:

Service A: label `important:true`, `notsoimportantbutimportant:true`
Service B: `notsoimportantbutimportant:true`
Service C

pool A -> selector: `important:true` , prio 100
pool B -> selector `notsoimportantbutimportant:true` (which includes some services collected by poolA), prio 50
pool C -> no selector

Reply:
Ohh, thanks, that is what I was missing!

Just curious, is that a real use case or just what happens with all the other stuff? I mean, it seems intricate if it is an use case, I'm just curious.

It seems reasonable. Also, if we want the alternative "to allocate from any of poolA or pool B", then pool B needs to have prio 100 (or pool A and pool B no prio).

I guess what we can't express is "take IPs from this pool, only when that pool is empty use this other pool". And that seems like a reasonable thing to leave out, AFAIK. Either services can have more labels (to play with selectors) and the user can request an specific IP.

Thanks for the explanations!

Reply:
The real (and most common) use case is "I want to reserve some pools for some namespaces / services", and other pools to the rest.
The extra complication comes from the fact that we can have intersections in those selections, and if we allow them, it will be possible to incur in what I described.

In any case, documenting the priority and saying that we'll go on the next pool if the highest one is empty (instead of failing) is as much as reasonable, and I think it makes sense.

Reply:
@rata changed.